packages/flutter/lib/src/widgets/slotted_render_object_widget.dart

```diff
@@ -231,14 +241,43 @@ class SlottedRenderObjectElement<S> extends RenderObjectElement {
     }(), '${widget.runtimeType}.slots must not change.');
     assert(slottedMultiChildRenderObjectWidgetMixin.slots.toSet().length == slottedMultiChildRenderObjectWidgetMixin.slots.length, 'slots must be unique');
 
+    final Map<Key, S> oldKeyedSlots = _keyToSlot;
+    _keyToSlot = <Key, S>{};
+
+    // New slots for keyed children. It shouldn't be updated in place because
+    // two children could swap slots.
+    final Map<S, Element> slotToKeyedChild = <S, Element>{};
+    final List<MapEntry<S, Widget?>> unkeyedWidgets = <MapEntry<S, Widget?>>[];
+
     for (final S slot in slottedMultiChildRenderObjectWidgetMixin.slots) {
-      _updateChild(slottedMultiChildRenderObjectWidgetMixin.childForSlot(slot), slot);
+      final Widget? widget = slottedMultiChildRenderObjectWidgetMixin.childForSlot(slot);
+      final Key? newWidgetKey = widget?.key;
+      if (newWidgetKey != null) {
+        assert(!_keyToSlot.containsKey(newWidgetKey));
+        _keyToSlot[newWidgetKey] = slot;
+        final S fromSlot = oldKeyedSlots[newWidgetKey] ?? slot;
+        final Element newChild = updateChild(_slotToChild.remove(fromSlot), widget, slot)!;
+        assert(!_slotToChild.containsKey(fromSlot));
+        assert(!slotToKeyedChild.containsKey(slot));
+        slotToKeyedChild[slot] = newChild;
+      } else {
+        // A unkeyed new widget shouldn't be used to update the child element
+        // yet, as a keyed widget that appears later in the iteration could
+        // claim that element.
+        unkeyedWidgets.add(MapEntry<S, Widget?>(slot, widget));
+      }
+    }
+
+    _slotToChild.addAll(slotToKeyedChild);
```

Comment:
Why do we need the indirection via `slotToKeyedChild`? Could these just be added to _slotToChild?